.github/workflows/rust.yml

```
@@ -13,39 +13,44 @@ env:
 
 jobs:
   lint:
-
     runs-on: ubuntu-latest
 
     steps:
-    - uses: actions/checkout@v2
-
-    - name: Install rust-toolchain
-      uses: actions-rs/toolchain@v1
-      with:
-        profile: minimal
-        components: rustfmt, clippy
-
-    - name: cargo fmt
-      uses: actions-rs/cargo@v1
-      with:
-        command: fmt
-        args: --all -- --check
-
-    - name: cargo clippy
-      uses: actions-rs/cargo@v1
-      env:
-        RUSTFLAGS: -D warnings
-      with:
-        command: clippy
-        args: --workspace --all-targets --all-features
+      - uses: actions/checkout@v2
+
+      - name: Install rust-toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          components: rustfmt, clippy
+
+      - name: cargo fmt
+        uses: actions-rs/cargo@v1
+        with:
+          command: fmt
+          args: --all -- --check
+
+      - name: cargo clippy
+        uses: actions-rs/cargo@v1
+        env:
+          RUSTFLAGS: -D warnings
+        with:
+          command: clippy
+          args: --workspace --all-targets --all-features
 
   build:
-
     runs-on: ubuntu-latest
 
     steps:
-    - uses: actions/checkout@v2
-    - name: Build (denying warnings)
-      run: RUSTFLAGS='-D warnings' cargo build --all-targets --all-features --verbose
-    - name: Run tests
-      run: cargo test --verbose
+      - uses: actions/checkout@v2
+      - name: Build (denying warnings)
+        run: RUSTFLAGS='-D warnings' cargo build --all-targets --all-features --verbose
+      - name: Run tests
+        run: cargo test --verbose
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+            target
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
```

```suggestion
          key: ${{ runner.os }}-cargo-${{ github.job }}-${{ hashFiles('**/Cargo.lock') }}
          restore-keys: |
            ${{ runner.os }}-cargo-${{ github.job }}-
            ${{ runner.os }}-cargo-
```

をおすすめします。

`restore-keys:` を書くと、「上から順番にキャッシュキーを先頭一致検索して、ヒットした場合はその名前のキャッシュキーに紐づくキャッシュをrestoreする」という動作をします。
`restore-keys:` を書かなければ `key:` に書いたキャッシュキーと完全一致検索をします。

この場合の問題点は、Cargo.lock が変わったらキャッシュを全く見なくなってしまうところです。更新されていないcrateもたくさんあるはずなのでキャッシュからrestoreしたいですね。

それと、キャッシュキーにはjob名を含めるようにしています。job名が含まれないと、「lint jobとbuild jobでできあがった `~/.cargo/registry` と `~/.cargo/git` を両方のjobで奪い合う形でsaveする」という挙動になります。もっというと、 `actions/cache@v2` は `key:` で指定したキーのキャッシュが既に存在する場合、キャッシュの保存を諦める（先勝ち）の方式をとっています。
実際に起こる問題としては、「lint jobのほうが先に完了してキャッシュをするが、lint jobでは引っ張ってくるcrateが少なかった。たくさんのcrateを引っ張ってきたbuild jobはもう後負けになって、キャッシュできなかった」結果、次回走るbuild jobでrestoreしたキャッシュには、build jobが必要とするcrateがあまり入っていない、ということがあります（実際これやらかしたことあります）。

`restore-keys:` の2つ目は、「restoreのときは、job名が同じキャッシュがなければ違うのからとっても良いよ」と指定しています。
健全に全jobが走っているときは恩恵がないのですが、例えば「buildジョブだけ落ち続けているのでbuildジョブのキャッシュはまだできていない」みたいなケースで、せめてlintジョブのキャッシュは使えるようになるという旨味があります。